core/src/main/scala/org/apache/spark/executor/ExecutorURLClassLoader.scala

```diff
@@ -32,36 +35,40 @@ private[spark] trait MutableURLClassLoader extends ClassLoader {
 }
 
 private[spark] class ChildExecutorURLClassLoader(urls: Array[URL], parent: ClassLoader)
-  extends MutableURLClassLoader {
-
-  private object userClassLoader extends URLClassLoader(urls, null){
-    override def addURL(url: URL) {
-      super.addURL(url)
-    }
-    override def findClass(name: String): Class[_] = {
-      super.findClass(name)
-    }
-  }
+  extends URLClassLoader(urls, null) with MutableURLClassLoader {
 
   private val parentClassLoader = new ParentClassLoader(parent)
 
-  override def findClass(name: String): Class[_] = {
+  override def loadClass(name: String, resolve: Boolean): Class[_] = {
     try {
-      userClassLoader.findClass(name)
```

Comment:
BTW, based on your comment (the link to the IBM article) on the other PR, it seems it would be safer to make this method synchronized just to be safe. I'd be surprised if any code in Spark would trigger a deadlock (it would take some really sketchy code in a class's static initializer to trigger that), but better to be safe.